File: option2.js
import React, { useState, useEffect } from 'react';
import axios from 'axios';

function Backtester() {
  const [strategy, setStrategy] = useState('moving_average');
  const [portfolio, setPortfolio] = useState('AAPL');
  const [start_date, setStartDate] = useState('2020-01-01');
  const [end_date, setEndDate] = useState('2020-12-31');
  const [results, setResults] = useState([]);

  useEffect(() => {
    const fetchData = async () => {
      const response = await axios.post('http://localhost:5000/backtest', {
        strategy,
        portfolio,
        start_date,
        end_date,
      });
      setResults(response.data.cumulative_returns);
    };
    fetchData();
  }, [strategy, portfolio, start_date, end_date]);

  return (
    <div>
      <h1>Backtester</h1>
      <form>
        <label>
          Strategy:
          <select value={strategy} onChange={(e) => setStrategy(e.target.value)}>
            <option value="moving_average">Moving Average</option>
            <option value="mean_reversion">Mean Reversion</option>
          </select>
        </label>
        <br />
        <label>
          Portfolio:
          <input type="text" value={portfolio} onChange={(e) => setPortfolio(e.target.value)} />
        </label>
        <br />
        <label>
          Start Date:
          <input type="date" value={start_date} onChange={(e) => setStartDate(e.target.value)} />
        </label>
        <br />
        <label>
          End Date:
          <input type="date" value={end_date} onChange={(e) => setEndDate(e.target.value)} />
        </label>
      </form>
      <h2>Results</h2>
      <ul>
        {results.map((result, index) => (
          <li key={index}>{result}</li>
        ))}
      </ul>
    </div>
  );
}

export default Backtester;

File: option2.py
from flask import Flask, request, jsonify
import pandas as pd
import yfinance as yf

app = Flask(__name__)

@app.route('/backtest', methods=['POST'])
def backtest():
    data = request.json
    strategy = data['strategy']
    portfolio = data['portfolio']
    start_date = data['start_date']
    end_date = data['end_date']

    # Load historical data for the portfolio
    portfolio_data = yf.download(portfolio, start=start_date, end=end_date)['Adj Close']

    # Apply the strategy to the portfolio data
    if strategy == 'moving_average':
        signals = pd.DataFrame(index=portfolio_data.index)
        signals['signal'] = 0.0
        signals['signal'][10:] = np.where(portfolio_data[10:] > portfolio_data[:-10].rolling(window=10).mean(), 1.0, 0.0)
        signals['positions'] = signals['signal'].diff()
    elif strategy == 'mean_reversion':
        signals = pd.DataFrame(index=portfolio_data.index)
        signals['signal'] = 0.0
        signals['signal'][10:] = np.where(portfolio_data[10:] < portfolio_data[:-10].rolling(window=10).mean(), -1.0, 0.0)
        signals['positions'] = signals['signal'].diff()

    # Calculate the returns of the strategy
    portfolio_returns = portfolio_data.pct_change()
    strategy_returns = portfolio_returns * signals['positions'].shift(1)

    # Calculate the cumulative returns of the strategy
    cumulative_returns = (1 + strategy_returns).cumprod()

    # Return the results
    return jsonify({'cumulative_returns': cumulative_returns.tolist()})

if __name__ == '__main__':
    app.run(debug=True)
